applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp

```diff
@@ -0,0 +1,1007 @@
+    mpMatrixConstitutiveLaw->CalculateMaterialResponseCauchy(values_matrix);
+    matrix_tangent_tensor = values_matrix.GetConstitutiveMatrix();
+
+    // Compute the tangent tensor of the fiber
+    values_fiber.SetMaterialProperties(r_props_fiber_cl);
+    mpFiberConstitutiveLaw->CalculateMaterialResponseCauchy(values_fiber);
+    fiber_tangent_tensor = values_fiber.GetConstitutiveMatrix();
+
+    noalias(matrix_tangent_tensor_ss) = prod(rSerialProjector, Matrix(prod(matrix_tangent_tensor,trans(rSerialProjector))));
+    noalias(fiber_tangent_tensor_ss)  = prod(rSerialProjector, Matrix(prod(fiber_tangent_tensor, trans(rSerialProjector))));
+
+    const double constant = (1.0 - mFiberVolumetricParticipation) / mFiberVolumetricParticipation;
+    Matrix jacobian_matrix(num_serial_components, num_serial_components);
+    noalias(jacobian_matrix) = matrix_tangent_tensor_ss + constant * fiber_tangent_tensor_ss;
+	Matrix inv_jacobian(num_serial_components, num_serial_components);
+    double det_jacobian;
+
+    MathUtils<double>::InvertMatrix(jacobian_matrix, inv_jacobian, det_jacobian);
+
+    rSerialStrainMatrix = rSerialStrainMatrix - prod(inv_jacobian, rResidualStresses);
+
+    // Previous flags restored
+    r_flags.Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN, flag_strain);
+    r_flags.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, flag_const_tensor);
+    r_flags.Set(ConstitutiveLaw::COMPUTE_STRESS, flag_stress);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+void SerialParallelRuleOfMixturesLaw::CheckStressEquilibrium(
+    const Vector& rStrainVector,
+    const Matrix& rSerialProjector,
+    const Vector& rMatrixStressVector,
+    const Vector& rFiberStressVector,
+    Vector& rStressSerialResidual,
+    bool& rIsConverged,
+    const Matrix& rConstitutiveTensorMatrixSS,
+    const Matrix& rConstitutiveTensorFiberSS
+)
+{
+    const Vector& r_serial_total_strain  = prod(rSerialProjector, rStrainVector);
+    const Vector& r_serial_stress_matrix = prod(rSerialProjector, rMatrixStressVector);
+    const Vector& r_serial_stress_fiber  = prod(rSerialProjector, rFiberStressVector);
+
+    const double norm_serial_stress_matrix = MathUtils<double>::Norm(r_serial_stress_matrix);
+    const double norm_serial_stress_fiber  = MathUtils<double>::Norm(r_serial_stress_fiber);
+    double ref = std::min(norm_serial_stress_matrix, norm_serial_stress_fiber);
+
+    // Here we compute the tolerance
+    double tolerance;
+    if (ref <= machine_tolerance) {
+        const double norm_product_matrix = MathUtils<double>::Norm(prod(rConstitutiveTensorMatrixSS, r_serial_total_strain));
+        const double norm_product_fiber  = MathUtils<double>::Norm(prod(rConstitutiveTensorFiberSS, r_serial_total_strain));
+        ref = std::min(norm_product_matrix, norm_product_fiber);
+    }
+    if (ref < 1e-9) tolerance = 1e-9;
```

Comment:
Please do not hardcode this, define a variable and add it to properties. take this value in case the property is not defined

Reply:
done